dap/debug_shell.go

```diff
@@ -0,0 +1,211 @@
+	sem *semaphore.Weighted
+
+	connected chan struct{}
+	cancel    context.CancelCauseFunc
+}
+
+func newShell() *shell {
+	return &shell{
+		fwd:       ioset.NewForwarder(),
+		sem:       semaphore.NewWeighted(1),
+		connected: make(chan struct{}),
+	}
+}
+
+// Init initializes the shell for connections on the client side.
+// Attach will block until the terminal has been initialized.
+func (s *shell) Init() error {
+	return s.listen()
+}
+
+func (s *shell) listen() error {
+	s.once.Do(func() {
+		var dir string
+		dir, s.err = os.MkdirTemp("", "buildx-dap-exec")
+		if s.err != nil {
+			return
+		}
+		defer func() {
+			if s.err != nil {
+				os.RemoveAll(dir)
+			}
+		}()
+		s.SocketPath = filepath.Join(dir, "s.sock")
+
+		s.l, s.err = net.Listen("unix", s.SocketPath)
+		if s.err != nil {
+			return
+		}
+
+		s.eg, _ = errgroup.WithContext(context.Background())
+		s.eg.Go(func() error {
+			conn, err := s.l.Accept()
+			if err != nil {
+				return err
+			}
+			fmt.Fprintln(conn, "Attached to build process.")
+
+			// Set the input of the forwarder to the connection.
+			s.fwd.SetIn(&ioset.In{
+				Stdin:  io.NopCloser(conn),
+				Stdout: conn,
+				Stderr: nopCloser{conn},
+			})
+			close(s.connected)
+
+			// Start a background goroutine to politely refuse any subsequent connections.
+			for {
+				conn, err := s.l.Accept()
+				if err != nil {
+					return nil
+				}
+				fmt.Fprint(conn, "Error: Already connected to exec instance.")
+				conn.Close()
+			}
+		})
+	})
+	return s.err
+}
+
+// Attach will attach the given build container to the shell.
+// Only one container can attach to a shell at any given time.
+// Other attaches will block until the context is canceled or it is
+// able to reserve the shell for its own use.
+//
+// This method is intended to be called by paused threads.
+func (s *shell) Attach(ctx context.Context, rCtx *build.ResultHandle) error {
+	cfg := &build.InvokeConfig{Tty: true}
+	if len(cfg.Entrypoint) == 0 && len(cfg.Cmd) == 0 {
+		cfg.Entrypoint = []string{"/bin/sh"} // launch shell by default
+		cfg.Cmd = []string{}
+		cfg.NoCmd = false
+	}
+	return s.attach(ctx, rCtx, cfg)
+}
+
+func (s *shell) attach(ctx context.Context, rCtx *build.ResultHandle, cfg *build.InvokeConfig) (retErr error) {
+	select {
+	case <-s.connected:
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+
+	in, out := ioset.Pipe()
+	defer in.Close()
+	defer out.Close()
+
+	s.fwd.SetOut(&out)
+	defer s.fwd.SetOut(nil)
+
+	// Check if the entrypoint is executable. If it isn't, don't bother
+	// trying to invoke.
+	if !s.canInvoke(ctx, rCtx, cfg) {
+		fmt.Fprintln(in.Stdout, "Waiting for build container...")
+		return nil
+	}
+
+	if err := s.sem.Acquire(ctx, 1); err != nil {
+		return err
+	}
+	defer s.sem.Release(1)
+
+	ctr, err := build.NewContainer(ctx, rCtx, cfg)
+	if err != nil {
+		return err
+	}
+	defer ctr.Cancel()
+
+	fmt.Fprintf(in.Stdout, "Running %s in build container.\n", strings.Join(append(cfg.Entrypoint, cfg.Cmd...), " "))
```

Comment:
```
Running /bin/sh in build container.
/ # 
Running /bin/sh in build container.
/ # 
Running /bin/sh in build container.
/ # 
```
Is there a way for us to include the line number or some other additional context information here I wonder? 🤔 For someone new to the feature they may think it is misbehaving if they see the same log message being sent multiple times.

Reply:
And if it's the default shell then just the following would be fine:

```
Running shell from "RUN echo foo"
/ # 
Running shell from "RUN echo bar"
/ # 
```

Reply:
> to include the line number

Or maybe line number is better as a similar command could exist in multiple places or could be very long in some Dockerfiles.

Reply:
I've included the line number and an additional message to say the changes will be reset when the build continues.